core/src/Kokkos_MathematicalSpecialFunctions.hpp

```diff
@@ -1268,6 +1268,222 @@ KOKKOS_INLINE_FUNCTION CmplxType cyl_bessel_h21(const CmplxType& z) {
   return ch21;
 }
 
+template <auto BesselZero, auto BesselOne, class CmplxType, class IntType,
+          class... Args>
+KOKKOS_INLINE_FUNCTION CmplxType cyl_bessel(const IntType& a,
+                                            const CmplxType& z, Args... args) {
+  if (a == 0) return BesselZero(z, args...);
+  if (a == 1) return BesselOne(z, args...);
+  CmplxType bessel_im2 = BesselZero(z, args...);
+  CmplxType bessel_im1 = BesselOne(z, args...);
+
+  for (IntType cur_a = 2; cur_a <= a; ++cur_a) {
+    CmplxType tmp = bessel_im1;
+    bessel_im1    = CmplxType(2 * (cur_a - 1)) * bessel_im1 / z - bessel_im2;
+    bessel_im2    = tmp;
+  }
+
+  return bessel_im1;
+}
+
+template <class CmplxType, class RealType, class IntType>
+KOKKOS_INLINE_FUNCTION CmplxType cyl_bessel_j(const IntType& a,
+                                              const CmplxType& z,
+                                              const RealType& joint_val = 25.0,
+                                              const IntType& bw_start   = 70) {
```

Comment:
I would rather not have these extra arguments (and remove the from the current public interface as well). These are implementation details.